mir_eval/transcription.py

```diff
@@ -0,0 +1,376 @@
+    est_pitches : np.ndarray, shape=(m,)
+        Array of estimated pitch values in Hertz
+    onset_tolerance : float > 0
+        The tolerance for an estimated note's onset deviating from the
+        reference note's onset, in seconds. Default is 0.05 (50 ms).
+    pitch_tolerance: float > 0
+        The tolerance for an estimated note's pitch deviating from the
+        reference note's pitch, in cents. Default is 50.0 (50 cents).
+    offset_ratio: float > 0 or None
+        The ratio of the reference note's duration used to define the
+        offset_tolerance. Default is 0.2 (20%), meaning the offset_tolerance
+        will equal the ref_duration * 0.2 * 0.5 (0.5 since the window is
+        centered on the reference offset), or 0.05 (50 ms), whichever is
+        greater. If ``offset_ratio`` is set to ``None``, offsets are ignored in
+        the matching.
+    offset_min_tolerance: float > 0
+        The minimum tolerance for offset matching. See offset_ratio description
+        for an explanation of how the offset tolerance is determined. Note:
+        this parameter only influences the results if ``offset_ratio`` is not
+        ``None``.
+    strict: bool
+        If ``strict=False`` (the default), threshold checks for onset, offset,
+        and pitch matching are performed using ``<=`` (less than or equal). If
+        ``strict=True``, the threshold checks are performed using ``<`` (less
+        than).
+
+    Returns
+    -------
+    matching : list of tuples
+        A list of matched reference and estimated notes.
+        ``matching[i] == (i, j)`` where reference note i matches estimate note
+        j.
+    """
+    # set the comparison function
+    if strict:
+        cmp = np.less
+    else:
+        cmp = np.less_equal
+
+    # check for onset matches
+    onset_distances = np.abs(np.subtract.outer(ref_intervals[:, 0],
+                                               est_intervals[:, 0]))
+    onset_hit_matrix = cmp(onset_distances, onset_tolerance)
+
+    # check for pitch matches
+    pitch_distances = np.abs(1200*np.log2(np.divide.outer(ref_pitches,
+                                                          est_pitches)))
+    pitch_hit_matrix = cmp(pitch_distances, pitch_tolerance)
+
+    # check for offset matches if offset_ratio is not None
+    if offset_ratio is not None:
+        offset_distances = np.abs(np.subtract.outer(ref_intervals[:, 1],
+                                                    est_intervals[:, 1]))
+        ref_durations = util.intervals_to_durations(ref_intervals)
+        offset_tolerances = 0.5 * offset_ratio * ref_durations
+        offset_tolerance_matrix = np.tile(offset_tolerances,
+                                          (offset_distances.shape[1], 1)).T
+        min_tolerance_inds = offset_tolerance_matrix < offset_min_tolerance
+        offset_tolerance_matrix[min_tolerance_inds] = offset_min_tolerance
+        offset_hit_matrix = cmp(offset_distances, offset_tolerance_matrix)
```

Comment:
You can do the same thing (I believe, please double check) more efficiently like so:

```Python
offset_tolerances = 0.5 * offset_ratio * ref_durations
min_tolerance_inds = offset_tolerances < offset_min_tolerance
offset_tolerances[min_tolerance_inds] = offset_min_tolerance
offset_hit_matrix = cmp(offset_distances, offset_tolerances.reshape(-1, 1))
```